### PR TITLE
Share build command flag configuration on desktop

### LINKED ISSUE #46236: Run `flutter pub get` when running `flutter build macos`
This happens in the https://github.com/flutter/plugins/tree/master/packages/connectivity/example and e2e plugin (the only macos plugins so far). This might be a problem with how they are setup, or a tooling problem.

### LINKED ISSUE #64944: Please support --dart-define for flutter build macos
## Use case

We would like to pass environment variables to a build process for macOS.
`flutter build` supports  passing environment variables to the build process when building for ios, android, windows and the web:

`flutter build web --dart-define=MY_HOSTNAME=api.example.com`

Unfortunately the option `--dart-define` is not available for macOS!

## Proposal

I noticed that for the web the minified main.dart.js contains the value passed with `--dart-define`.
It would be great, if the same would be done for macOS, leading to modified compiled native app including the passed value.

### PR DESCRIPTION
**Pull request by stuartmorgan-g**
## Description

Almost all of the flag setup on Linux, macOS, and Windows should be
identical; this pulls that common setup to a shared method.

This adds support for several flags on macOS that have all the necessary
plumbing already in place due to shared backend code, but were never
enabled.

## Related Issues

Fixes https://github.com/flutter/flutter/issues/64944
Fixes https://github.com/flutter/flutter/issues/46236

## Tests

I added the following tests: macOS version of the existing Windows/Linux tests that verify that the various flags result in the correct native build configurations.

## Checklist

Before you create this PR, confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter_tools/test/commands.shard/hermetic/build_macos_test.dart [hunk 1]
```diff
@@ -112,7 +112,7 @@ void main() {
     createCoreMockProjectFiles();
 
     expect(createTestCommandRunner(command).run(
-      const <String>['build', 'macos']
+      const <String>['build', 'macos', '--no-pub']
```
Comment (jonahwilliams) [inline packages/flutter_tools/test/commands.shard/hermetic/build_macos_test.dart:115]: Oh hey, this was another bug we needed to fix

**Review by jonahwilliams — APPROVED** [review]
LGTM

File: packages/flutter_tools/test/commands.shard/hermetic/build_macos_test.dart (hunk 1 above)
Comment (jonahwilliams) [reply packages/flutter_tools/test/commands.shard/hermetic/build_macos_test.dart:115]: I can't find the original issue, but we did not correctly run pub when building for macos

Comment (stuartmorgan) [reply packages/flutter_tools/test/commands.shard/hermetic/build_macos_test.dart:115]: https://github.com/flutter/flutter/issues/46236

**Pull request merged**